clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -83,13 +87,71 @@ void BoundsAnalysis::InitInOutSets() {
   }
 }
 
+llvm::APSInt BoundsAnalysis::GetSwitchCaseVal(const Expr *CaseExpr) {
+  Expr *E = const_cast<Expr *>(CaseExpr);
+
+  while (auto *ICE = dyn_cast<ImplicitCastExpr>(E))
+    E = ICE->getSubExpr();
+
+  // const int i = 1;
+  // case i: This is how it is represented in the CFG:
+  // ConstantExpr 'const int' lvalue 1
+  //   -DeclRefExpr 'const int' lvalue Var 'i' 'const int'
+  // This is not an IntegerConstantExpr. It is a ConstantExpr and we get its
+  // value by calling the getResultAsAPSInt() method.
+
+  // case '1': This is how it is represented in the CFG:
+  // ConstantExpr 'int'
+  //   -CharacterLiteral 'int' 49
+  // This is an IntegerConstantExpr. After the call to isIntegerConstantExpr,
+  // the variable IntVal would contain the value '1'.
+
+  llvm::APSInt IntVal (Ctx.getTypeSize(Ctx.IntTy), 0);
+
+  if (const auto *CE = dyn_cast_or_null<ConstantExpr>(E)) {
+    if (!E->isIntegerConstantExpr(IntVal, Ctx))
+      IntVal = CE->getResultAsAPSInt();
+  }
+  return IntVal;
+}
+
+bool BoundsAnalysis::CheckIsSwitchCaseNull(ElevatedCFGBlock *EB) {
+  if (const auto *CS = dyn_cast_or_null<CaseStmt>(EB->Block->getLabel())) {
+    llvm::APSInt Zero (Ctx.getTypeSize(Ctx.IntTy), 0);
```

Comment:
I believe this may not be the correct type for zero.   The C11 specification says in 6.8.4.2 that the type of the constant expression in a case expression shall be an integer constant expression.   There are a number of different sizes of integer types in C, which are distinct from the `int` type.  The C11 specification says that integer promotions are applied to the controlling expression in the switch statement, and that the label expressions of the case statements are promoted to match that type.   

I believe that the comparison functions for APSInt  may expect the integers to have the same size.   It's been a while since I used the API, so I amy be wrong. In that case, the test for inclusion of zero, as well as the uses of APSInt in GetSwitchCaseVal could both lead to internal compiler assertion failures, if the integer requires more bits than IntTy.

My suggestion is that you test out switches that use an `unsigned int`, as well as an integer type larger than the default size of `int`.  I believe the default promotions rules for `char` will widen everything out to `int`, which is why the code is working for your test cases.

Reply:
I have added new tests for unsigned, UINT_MAX, INT_MAX and INT_MIN case labels. The compiler and GetSwitchCaseVal seem to handle these correctly. For example:

`case 999999999999999999999999999` is internally represented as `11515845246265065471ULL` and we return detect that this is non-zero.

In case of `case INT_MIN + INT_MIN` we do not widen as the compiler evaluates this to 0 with the warning: "overflow in expression; result is 0 with type 'int'".